### PR TITLE
[WIP] Improve AfformSubmission creation

### PR DESCRIPTION
**Pull request by MegaphoneJon**
Overview
----------------------------------------
AfformSubmission is a necessary improvement, but they aren't recorded if there's an error in the form submission, nor do they capture enough information for troubleshooting.  With this PR, we should be able to replay Form Builder submissions, both for troubleshooting and to replicate Webform "resubmit" functionality.

Before
----------------------------------------
* AfformSubmission created after the Afform is processed (despite not relying on any processing code).
* AfformSubmission only contains entity IDs.

After
----------------------------------------
* AfformSubmission created before the Afform is processed
* AfformSubmission contains all submitted values.

Comments
----------------------------------------
This is especially important right now because Form Builder is failing silently when it can't process data (e.g. field not filled in).

### CODE REVIEW

**Comment by colemanw** [comment]
This is a good improvement except we'll lose the IDs and those are useful especially if you want to go back and edit the form submission (when we add that feature). Maybe we can store both.

**Comment by MegaphoneJon** [comment]
I was just now wondering why we were storing the IDs, but that's an excellent point.  Yeah, let's store both, I'll update this.

**Comment by MegaphoneJon** [comment]
Incorporated Coleman's feedback.

**Comment by colemanw** [comment]
I think this is approximately right but I'd like to see about combining the `ids` array with the `values` array because I think they have the same structure and could be recursively merged.

**Comment by aydun** [comment]
See also https://lab.civicrm.org/dev/core/-/issues/3744

**Comment by MegaphoneJon** [comment]
@colemanw I took a crack at combining the arrays but the structure isn't the same - though I'm not sure if that's intentional or not.

Here's `$this->_entityIds` before my patch on a basic Afform with joins:
Selection_1600

Note that the entity ID for organization1 is `$this->_entityIds['Organization1'][0]['id']` but the email id is:
`$this->_entityIds['Organization1'][0]['joins']['Email'][0]`.

To be recursively mergeable, the Email ID should be at:
`$this->_entityIds['Organization1'][0]['joins']['Email'][0]['id]`...or we need to add special handling for joins.

I'll push the code I wrote to look at, but that officially makes this WIP.

**Comment by colemanw** [comment]
I see. Let me see if I can cleanup that structure as it's mostly internal.

**Comment by colemanw** [comment]
While stepping through that code I discovered a rather major bug in how joins are handled - they have a bad habit of deleting existing records (due to the use of the api `replace` action) which will become more problematic when we add the existing contact selection. So I'm going to work on a PR to fix that while also cleaning up the `$entityIds` structure.

**Comment by colemanw** [comment]
@MegaphoneJon I believe https://github.com/civicrm/civicrm-core/pull/24264 will make this one mergeable

File: ext/afform/core/Civi/Api4/Action/Afform/Submit.php [hunk 1]
```diff
@@ -82,6 +91,21 @@ protected function processForm() {
     ];
   }
```
Comment (briennekordis) [inline ext/afform/core/Civi/Api4/Action/Afform/Submit.php:93]: /**
   * Recursively removes 'joins' from $submissionData in favor of submiting only '_joins'
   */
  private function removeJoinsElement(&$submissionData) {
    foreach ($submissionData as $key => &$value) {
      if ($key === 'joins') {
        unset($submissionData[$key]);
      }
      elseif (is_array($value)) {
        $this->removeJoinsElement($value);
      }
    }
    return $submissionData;
  }

Comment (briennekordis) [inline ext/afform/core/Civi/Api4/Action/Afform/Submit.php:93]:
```suggestion
  /**
   * Recursively removes 'joins' from $submissionData in favor of submiting only '_joins'
   */
  private function removeJoinsElement(&$submissionData) {
    foreach ($submissionData as $key => &$value) {
      if ($key === 'joins') {
        unset($submissionData[$key]);
      }
      elseif (is_array($value)) {
        $this->removeJoinsElement($value);
      }
    }
    return $submissionData;
  }

```

File: ext/afform/core/Civi/Api4/Action/Afform/Submit.php [hunk 2]
```diff
@@ -67,13 +76,13 @@ protected function processForm() {
       \Civi::dispatcher()->dispatch('civi.afform.submit', $event);
     }
 
-    // Save submission record
+    $submissionData = $this->array_insert($this->getValues(), $this->_entityIds);
```
Comment (briennekordis) [inline ext/afform/core/Civi/Api4/Action/Afform/Submit.php:79]:
```suggestion

```

File: ext/afform/core/Civi/Api4/Action/Afform/Submit.php [hunk 3]
```diff
@@ -67,13 +76,13 @@ protected function processForm() {
...
+    // Update submission record with entity IDs.
     if (!empty($this->_afform['create_submission'])) {
-      $submission = AfformSubmission::create(FALSE)
-        ->addValue('contact_id', \CRM_Core_Session::getLoggedInContactID())
-        ->addValue('afform_name', $this->name)
-        ->addValue('data', $this->_entityIds)
-        ->execute()->first();
+      AfformSubmission::update(FALSE)
```
Comment (briennekordis) [inline ext/afform/core/Civi/Api4/Action/Afform/Submit.php:82]:
```suggestion
    $submissionData = $this->array_insert($this->getValues(), $this->_entityIds);
    $this->removeJoinsElement($submissionData);
    AfformSubmission::update(FALSE)
```

Comment (briennekordis) [inline ext/afform/core/Civi/Api4/Action/Afform/Submit.php:82]:
```suggestion
      $submissionData = $this->array_insert($this->getValues(), $this->_entityIds);
      $this->removeJoinsElement($submissionData);
      AfformSubmission::update(FALSE)
```

**Comment by colemanw** [comment]
I've combined this into #24264

**Pull request closed without merging**